### DISCUSSION TITLE
Provide a standardised log system ?

### DISCUSSION THREAD

**Discussion by cprudhom** [chocoteam/choco-solver#1032] [Ideas]
I wonder if it wouldn't be a good idea to use a log system, like SLF4J. To date, choco has a dummy log system, which allows you to display certain information about the model or the resolution (I think it's the latter that's used).
Mainly, they are defined in `IOutputFactory.java`.

At the moment, it's pretty easy to display in the console (via `System.out`) the characteristics of the model, the resolution statistics, the decisions made, the solutions found and the failures encountered...

However, it might be useful to plot more information, such as that relating to modelling or resolution configuration.
For instance, it's not always easy to know if a constraint has not been posted...
The difficulty is determining what should appear in the logs, what should appear in the console (by default), what can be activated/deactivated, etc.

But first of all, does such a need exist, or is what is there enough?

**Comment by jgFages** [comment]
I already asked myself the same question. 
I usually use the prints of choco for debugging / testing only. 
In production, I use my own logs that indeed rely on a proper logging system. 
Having a logging system in choco solver might be a plus, but I am not sure it is worth it. 
I am happy with what is already in place.

**Comment by martins-avots** [comment]
If there are no special requirements for the logging,
Java's internal logging framework JUL (`java.util.logging`) might be interesting,
as no new dependencies are required in this case.
Bridges for SLF4J and Log4J2 seem to exist as well.

Also, I have to admit, that I have no personal experience with JUL.
mhh, people are not that happy about it, so maybe JUL is not that great: https://stackoverflow.com/questions/11359187/why-not-use-java-util-logging

**Comment by cprudhom** [comment]
Thank you for your feedbacks.

We agree that this is more of a development aid than a production aid.
At the moment, it's pretty easy to get simple information about the model (number of variables, constraints) and the solution (number of solutions, failures, value of the objective function, etc.).  

To me the questions are: 
Is there a need to access to other elements? Which ones?
Is there a need to formalise this information?
Basically, is there a need to extract knowledge about models and their resolution?

**Reply by jgFages** [reply]
Not on my side
